tutorials/text/deduplication/semantic/semantic_step_by_step.ipynb

```diff
@@ -0,0 +1,1153 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "54c8fe16",
+   "metadata": {},
+   "source": [
+    "# Step by Step Semantic Deduplication on Text Data\n",
+    "\n",
+    "GPU accelerated implementation of [SemDeDup: Data-efficient learning at web-scale through semantic deduplication](https://arxiv.org/abs/2303.09540)  \n",
+    "\n",
+    "The tutorial here shows how to run Semantic Duplication on text data by executing three workflows sequentially. \n",
+    "\n",
+    "We also use ID generator to show how it works when running it separately.\n",
+    "\n",
+    "1. Create ID generator.\n",
+    "2. Running embedding generation\n",
+    "3. Running kmeans + pairwise (without duplicate identification)\n",
+    "4. Run duplicate identification\n",
+    "5. Run removal\n",
+    "\n",
+    "We also allow users to run these steps as a single workflow, which can be seen in end to end tutorial in the same directory as this tutorial."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "8a2c97d2",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import os\n",
+    "\n",
+    "os.environ[\"LOGURU_LEVEL\"] = \"ERROR\"\n",
+    "\n",
+    "import pandas as pd\n",
+    "\n",
+    "input_path = os.path.abspath(\"./input\")\n",
+    "semantic_out_dir = os.path.abspath(\"./output/step_by_step\")\n",
+    "output_path = os.path.join(semantic_out_dir, \"output\")\n",
+    "cache_path = os.path.join(semantic_out_dir, \"cache\")\n",
+    "# this can be either of jsonl or parquet\n",
+    "# #however if you change it to jsonl you'll need to change embedding generations reader to also be JsonlReader\n",
```

Comment:
```suggestion
```
Instead of this comment you could do something like
```
if input_filetype == "parquet":
    reader_stage = ParquetReader(file_paths=input_path, files_per_partition=1, fields=["text"], _generate_ids=True)
elif input_filetype == "jsonl":
    reader_stage = JsonlReader(...)
```

Reply:
I'd rather keep it simple in the subsequent stages otherwise I'm just reimplementing the workflow (which handles the if / else branching)